src/runtime/runtime_internal.h

```diff
@@ -30,8 +30,21 @@ typedef __PTRDIFF_TYPE__ ptrdiff_t;
 typedef ptrdiff_t ssize_t;
 
 #define NULL 0
+
+// In Halide runtime code, WEAK should be used almostly exclusively for
+// functions that are part of the public Halide runtime API.
 #define WEAK __attribute__((weak))
 
+// In Halide runtime code, helper functions that are internal to the runtime
```

Comment:
This comment isn't quite right - most functions should just be weak, whether or not they're part of the public API. always_inline is for things that either should be inlined for performance reasons, or for things that go into every compiled pipeline (not just the standalone runtime). Those things have to either disappear entirely by being inlined away or have "inline" linkage to avoid multiple definition errors on platforms that have no weak linkage.

Reply:
PTAL